Adding Blog Functionality and New Blog Posts

QuillPusher opened the pull request:
- Preparing to add multiple Blog Posts (including Xeus-Cpp) that I've gathered info on. Will also share to LinkedIn.

- Added **blog.html** layout and **blog.md** page to add Blog functionality

- **Reformatted existing articles** to work with default blog functionality (file names, dates, excerpts, etc.)

- **moved articles from "_pages/blogs" to "_posts"** as per standard Jekyll practice (this helps generate default variables like post.url)

- moved articles still work with old link, so **linkbacks will not be broken** (tested)

   - example: https://compiler-research.org/blogs/gsod23_quillpusher_experience_blog/

- **Did not add "Blog" menu to Home page navigation**, will do that as part of Front Page Customization PR

- Note: I found that following images were missing on Fred's blog (cc1.gif, cc2.gif, tcc.gif). I have added screenshots from his presentation for now and emailed him for actual moving GIF files that show Code Completion demo

QuillPusher commented:
Added **5 more Blog Posts** 

**Migrated from LLVM Blog (credit to original post added):**
(Converted Hugo-specific markdown to Liquid)

- Interactive C++ with Cling

- Interactive C++ for Data Science

- Interactive programming in C++ - 2022 GSoC internships

- Another step forward towards interactive programming

**New** 🆕 🥳  :

- Xeus Cpp: C++ and Python Integration in a Single Jupyter Environment

**Blog Post View**

<img width="947" alt="Blog_sample2" src="https://github.com/compiler-research/compiler-research.github.io/assets/130300172/6d0af537-42a9-4b87-a55c-25b0358285e9">

QuillPusher commented:
There is still room to add more blog posts over time, based on topics like 
- "How to get started with AD", 
- "Adopting Execution Result Handling Features", 
- "Interoperability between Cppyy and Numba", etc.
and these can be converted so social media [posts on LinkedIn](https://www.linkedin.com/feed/update/urn:li:activity:7172536612348751872?utm_source=share&utm_medium=member_desktop) and published on a delayed schedule. Example:

![BLOG](https://github.com/compiler-research/compiler-research.github.io/assets/130300172/b938dc57-5fa7-43ba-949d-6eb7cff92fc0)

davidlange6 commented:
eg they are here already, no
 https://github.com/compiler-research/compiler-research.github.io/tree/master/images/blog

QuillPusher commented:
Nice catch, I hadn't noticed these, updating shortly to use existing images

QuillPusher commented:
just found that the previous images are from a "Blog Draft" page that is not rendered anywhere on the website.

https://github.com/compiler-research/compiler-research.github.io/commit/4f92a925ef93528c7cafe6377ec81be4e6ddbe84#diff-f97301cb74fc1a8f69b28c9b1112daa77c8ec91d44d2ba2429a63420348ef279

QuillPusher commented:
I suggest that we let my changes go in as they are, since the image files are named logically, based on blog post name. Also I formatted the images to look better with the rest of the content.

The existing images and the "_pages/blog_post.md" can perhaps be removed in a separate PR, along with other dead areas of the website that are not rendered anywhere

vgvassilev commented:
That makes the repository bigger and the deployment of the website unnecessarily slower. Even if we delete these files afterwords they still contribute to the overall size as they are present in the repository history.

davidlange6 commented:
and here is where they are used on the website...
https://compiler-research.org/interactive_cpp
best not to duplicate content

QuillPusher commented:
@davidlange6 

The plot thickens 😄 

the link https://compiler-research.org/interactive_cpp takes the image from:

https://blog.llvm.org/img/cling-2020-11-30-figure1.gif

not from `/images/blog`

QuillPusher commented:
On a closer look the older images are inconsistent. 

- There is **one image missing** (cling-2020-12-21-figure3.png). 
- One of them is a **PNG instead of a GIF** (cling-2020-12-21-figure9.gif/figure10.png) which defeats the purpose
- There's another one that's **different from the original image** (cling-2020-12-21-figure1.png/figure3.png). 
- Also some of the older images are **lower resolution**. 

Please let me know if I should revert to older images, because then that would require me deleting some images and adding some images anyway.

vgvassilev commented:
> On a closer look the older images are inconsistent.
> 
>     * There is **one image missing** (cling-2020-12-21-figure3.png).

Adding this should be fine.

> 
>     * One of them is a **PNG instead of a GIF** (cling-2020-12-21-figure9.gif/figure10.png) which defeats the purpose

Those are two different items then and adding the gif is probably fine.

> 
>     * There's another one that's **different from the original image** (cling-2020-12-21-figure1.png/figure3.png).

Let's replace the wrong image.

> 
>     * Also some of the older images are **lower resolution**.

Let's replace the lower resolution images with the higher resolution ones.

> 
> 
> Please let me know if I should revert to older images, because then that would require me deleting some images and adding some images anyway.

We should update the ones that are crappy and add the ones that are not duplicated.

QuillPusher commented:
### Please ignore previous messages, full summary below: 🚀 🚀 🚀 

- created a `blog.html` Layout (used by blog.md)

- created a `blog.md` page (main blog page, where excerpts of blogs are listed)

- moved existing blog posts from `_pages\blogs` to `_posts` as per standard jekyll practice (this makes available several variables like `post.url`)

- added `date` parameter to existing blog posts' front matter to use for blog sorting. 

- Changed Layout to `post` for blog posts (uses the existing `post.html` layout. I have tested it, the output looks OK). 

- Added 'title' and 'excerpt' in front matter to display on main Blog page (otherwise they would have been auto created based on file name and first paragraph)

- removed additional posts and images (will be a separate PR)

- Will add a 'how to add a blog entry' guide separately as part of a new 'contribution guide'.

vgvassilev reviewed (commented):
This looks good to me.
